File: web/backend/services/mathComparisonGenerator.js
// services/mathComparisonGenerator.js

/**
 * Mathematical Comparison Puzzle Generator
 * Generates sequences of comparison puzzles with progressive difficulty
 */

export class MathComparisonGenerator {
    constructor() {
      this.debugMode = false;
    }
  
    setDebugMode(enabled) {
      this.debugMode = enabled;
    }
  
    debugLog(message, type = 'info') {
      if (this.debugMode) {
        const timestamp = new Date().toISOString();
        const emoji = type === 'error' ? '❌' : type === 'success' ? '✅' : type === 'warning' ? '⚠️' : '🔄';
        console.log(`${emoji} [${timestamp}] MATH_COMPARISON: ${message}`);
      }
    }
  
    /**
     * Generate a complete mathematical comparison puzzle
     */
    async generateMathComparisonPuzzle(difficulty = 'easy') {
      this.debugLog(`🎯 Generating ${difficulty} mathematical comparison puzzle...`);
  
      try {
        const difficultyConfig = this.getDifficultyConfig(difficulty);
        const puzzleSequence = this.generatePuzzleSequence(difficultyConfig);
        
        const questionData = {
          sequence: puzzleSequence,
          totalPairs: puzzleSequence.length,
          difficulty: difficulty,
          timeLimit: difficultyConfig.timeLimit,
          scoring: difficultyConfig.scoring,
          instructions: "Compare the values and select which is greater, or select EQUAL if they are the same.",
          metadata: {
            generatedAt: new Date().toISOString(),
            difficulty: difficulty,
            sequenceLength: puzzleSequence.length,
            operationTypes: this.getOperationTypes(puzzleSequence)
          }
        };
  
        const answerData = {
          correctAnswers: puzzleSequence.map(pair => pair.correctAnswer),
          scoring: difficultyConfig.scoring,
          maxScore: puzzleSequence.length * difficultyConfig.scoring.pointsPerCorrect
        };
  
        const result = {
          success: true,
          puzzleData: {
            question: JSON.stringify(questionData),
            answer: JSON.stringify(answerData),
            hint: `Compare ${puzzleSequence.length} mathematical expressions and choose the greater value or EQUAL`,
            difficulty: difficulty,
            metadata: {
              totalPairs: puzzleSequence.length,
              generatedAt: questionData.metadata.generatedAt,
              puzzleType: 'mathematical_comparison'
            }
          }
        };
  
        this.debugLog(`✅ Generated ${puzzleSequence.length} comparison pairs for ${difficulty} difficulty`, 'success');
        return result;
  
      } catch (error) {
        this.debugLog(`❌ Generation failed: ${error.message}`, 'error');
        return {
          success: false,
          message: `Math comparison generation failed: ${error.message}`
        };
      }
    }
  
    /**
     * Get difficulty configuration
     */
    getDifficultyConfig(difficulty) {
      const configs = {
        easy: {
          sequenceLength: 8,
          numberRange: [1, 20],
          operations: ['addition', 'subtraction', 'simple_multiplication'],
          allowDecimals: false,
          allowNegatives: false,
          timeLimit: 60000, // 60 seconds
          scoring: {
            pointsPerCorrect: 100,
            timeBonus: true,
            streakMultiplier: 1.1
          }
        },
        medium: {
          sequenceLength: 12,
          numberRange: [1, 50],
          operations: ['addition', 'subtraction', 'multiplication', 'division', 'mixed'],
          allowDecimals: true,
          allowNegatives: false,
          timeLimit: 90000, // 90 seconds
          scoring: {
            pointsPerCorrect: 150,
            timeBonus: true,
            streakMultiplier: 1.2
          }
        },
        hard: {
          sequenceLength: 15,
          numberRange: [1, 100],
          operations: ['multiplication', 'division', 'mixed', 'powers', 'fractions'],
          allowDecimals: true,
          allowNegatives: true,
          timeLimit: 120000, // 120 seconds
          scoring: {
            pointsPerCorrect: 200,
            timeBonus: true,
            streakMultiplier: 1.3
          }
        }
      };
  
      return configs[difficulty.toLowerCase()] || configs.medium;
    }
  
    /**
     * Generate a sequence of comparison pairs with progressive difficulty
     */
    generatePuzzleSequence(config) {
      const sequence = [];
      const { sequenceLength, operations, numberRange, allowDecimals, allowNegatives } = config;
  
      // Progressive difficulty within the sequence
      for (let i = 0; i < sequenceLength; i++) {
        const progressRatio = i / sequenceLength;
        const currentConfig = this.getProgressiveConfig(config, progressRatio);
        
        const pair = this.generateComparisonPair(currentConfig, i + 1);
        sequence.push(pair);
      }
  
      return sequence;
    }
  
    /**
     * Adjust difficulty progressively through the sequence
     */
    getProgressiveConfig(baseConfig, progressRatio) {
      const { numberRange, operations } = baseConfig;
      
      // Increase number range as we progress
      const rangeDiff = numberRange[1] - numberRange[0];
      const newMax = Math.round(numberRange[0] + rangeDiff * (0.3 + progressRatio * 0.7));
      
      return {
        ...baseConfig,
        numberRange: [numberRange[0], newMax],
        complexity: progressRatio
      };
    }
  
    /**
     * Generate a single comparison pair
     */
    generateComparisonPair(config, pairNumber) {
      const { operations, numberRange, allowDecimals, allowNegatives, complexity } = config;
      
      // Choose operation type based on difficulty progression
      const operationType = this.selectOperation(operations, complexity);
      
      // Generate the two values to compare
      const leftValue = this.generateValue(operationType, config, 'left');
      const rightValue = this.generateValue(operationType, config, 'right');
      
      // Calculate actual numeric values
      const leftNumeric = this.evaluateExpression(leftValue);
      const rightNumeric = this.evaluateExpression(rightValue);
      
      // Determine correct answer
      let correctAnswer;
      const tolerance = 0.001; // For floating point comparison
      
      if (Math.abs(leftNumeric - rightNumeric) < tolerance) {
        correctAnswer = 'equal';
      } else if (leftNumeric > rightNumeric) {
        correctAnswer = 'left';
      } else {
        correctAnswer = 'right';
      }
  
      return {
        pairNumber,
        leftValue,
        rightValue,
        leftNumeric,
        rightNumeric,
        correctAnswer,
        operationType,
        difficulty: this.calculatePairDifficulty(leftValue, rightValue, operationType)
      };
    }
  
    /**
     * Select operation type based on complexity
     */
    selectOperation(availableOps, complexity) {
      // Weight operations by complexity
      const operationWeights = {
        'addition': 0.1,
        'subtraction': 0.2,
        'simple_multiplication': 0.3,
        'multiplication': 0.5,
        'division': 0.7,
        'mixed': 0.8,
        'powers': 0.9,
        'fractions': 0.95
      };
  
      // Filter operations that are appropriate for current complexity
      const suitableOps = availableOps.filter(op => 
        operationWeights[op] <= (complexity + 0.3)
      );
  
      if (suitableOps.length === 0) {
        return availableOps[0]; // Fallback
      }
  
      return suitableOps[Math.floor(Math.random() * suitableOps.length)];
    }
  
    /**
     * Generate a mathematical value/expression
     */
    generateValue(operationType, config, side) {
      const { numberRange, allowDecimals } = config;
      
      switch (operationType) {
        case 'addition':
          return this.generateAddition(numberRange, allowDecimals);
        
        case 'subtraction':
          return this.generateSubtraction(numberRange, allowDecimals);
        
        case 'simple_multiplication':
          return this.generateSimpleMultiplication(numberRange);
        
        case 'multiplication':
          return this.generateMultiplication(numberRange, allowDecimals);
        
        case 'division':
          return this.generateDivision(numberRange, allowDecimals);
        
        case 'mixed':
          return this.generateMixedExpression(numberRange, allowDecimals);
        
        case 'powers':
          return this.generatePowerExpression(numberRange);
        
        case 'fractions':
          return this.generateFractionExpression(numberRange);
        
        default:
          return this.generateSimpleNumber(numberRange, allowDecimals);
      }
    }
  
    /**
     * Generate addition expression
     */
    generateAddition(range, allowDecimals) {
      const a = this.randomNumber(range, allowDecimals);
      const b = this.randomNumber(range, allowDecimals);
      return `${a} + ${b}`;
    }
  
    /**
     * Generate subtraction expression
     */
    generateSubtraction(range, allowDecimals) {
      const a = this.randomNumber(range, allowDecimals);
      const b = this.randomNumber([range[0], Math.min(a, range[1])], allowDecimals);
      return `${a} - ${b}`;
    }
  
    /**
     * Generate simple multiplication (single digits)
     */
    generateSimpleMultiplication(range) {
      const a = this.randomNumber([2, 9], false);
      const b = this.randomNumber([2, 9], false);
      return `${a} × ${b}`;
    }
  
    /**
     * Generate multiplication expression
     */
    generateMultiplication(range, allowDecimals) {
      // Keep numbers smaller for multiplication
      const maxVal = Math.min(range[1], 15);
      const a = this.randomNumber([range[0], maxVal], allowDecimals);
      const b = this.randomNumber([range[0], maxVal], allowDecimals);
      return `${a} × ${b}`;
    }
  
    /**
     * Generate division expression
     */
    generateDivision(range, allowDecimals) {
      const b = this.randomNumber([2, Math.min(range[1], 10)], false);
      const result = this.randomNumber([1, range[1]], allowDecimals);
      const a = b * result;
      return `${a} ÷ ${b}`;
    }
  
    /**
     * Generate mixed expression
     */
    generateMixedExpression(range, allowDecimals) {
      const operations = ['+', '-', '×'];
      const op1 = operations[Math.floor(Math.random() * operations.length)];
      const op2 = operations[Math.floor(Math.random() * operations.length)];
      
      const a = this.randomNumber([range[0], Math.min(range[1], 12)], allowDecimals);
      const b = this.randomNumber([range[0], Math.min(range[1], 12)], allowDecimals);
      const c = this.randomNumber([range[0], Math.min(range[1], 12)], allowDecimals);
      
      return `${a} ${op1} ${b} ${op2} ${c}`;
    }
  
    /**
     * Generate power expression
     */
    generatePowerExpression(range) {
      const base = this.randomNumber([2, Math.min(range[1], 8)], false);
      const exponent = this.randomNumber([2, 4], false);
      return `${base}^${exponent}`;
    }
  
    /**
     * Generate fraction expression
     */
    generateFractionExpression(range) {
      const numerator = this.randomNumber([1, range[1]], false);
      const denominator = this.randomNumber([2, 10], false);
      return `${numerator}/${denominator}`;
    }
  
    /**
     * Generate simple number
     */
    generateSimpleNumber(range, allowDecimals) {
      return this.randomNumber(range, allowDecimals).toString();
    }
  
    /**
     * Generate random number within range
     */
    randomNumber(range, allowDecimals) {
      const min = range[0];
      const max = range[1];
      
      if (allowDecimals && Math.random() < 0.3) { // 30% chance for decimals
        return Math.round((Math.random() * (max - min) + min) * 10) / 10;
      } else {
        return Math.floor(Math.random() * (max - min + 1)) + min;
      }
    }
  
    /**
     * Evaluate mathematical expression to numeric value
     */
    evaluateExpression(expression) {
      try {
        // Convert mathematical symbols to JavaScript operators
        let jsExpression = expression
          .replace(/×/g, '*')
          .replace(/÷/g, '/')
          .replace(/\^/g, '**')
          .replace(/(\d+)\/(\d+)/g, '($1/$2)'); // Handle fractions
        
        // Use Function constructor for safe evaluation (in production, use a proper math parser)
        return Function(`"use strict"; return (${jsExpression})`)();
      } catch (error) {
        this.debugLog(`Error evaluating expression "${expression}": ${error.message}`, 'error');
        return 0;
      }
    }
  
    /**
     * Calculate difficulty rating for a pair
     */
    calculatePairDifficulty(leftValue, rightValue, operationType) {
      let difficulty = 1;
      
      // Base difficulty by operation type
      const operationDifficulty = {
        'addition': 1,
        'subtraction': 1.2,
        'simple_multiplication': 1.5,
        'multiplication': 2,
        'division': 2.5,
        'mixed': 3,
        'powers': 3.5,
        'fractions': 4
      };
      
      difficulty *= operationDifficulty[operationType] || 1;
      
      // Increase difficulty for longer expressions
      const avgLength = (leftValue.length + rightValue.length) / 2;
      difficulty *= (1 + avgLength * 0.1);
      
      return Math.round(difficulty * 10) / 10; // Round to 1 decimal
    }
  
    /**
     * Get summary of operation types used
     */
    getOperationTypes(sequence) {
      const types = {};
      sequence.forEach(pair => {
        types[pair.operationType] = (types[pair.operationType] || 0) + 1;
      });
      return types;
    }
  
    /**
     * Format puzzle for storage (integrates with existing puzzle system)
     */
    formatForStorage(puzzleData, difficulty) {
      return {
        question: puzzleData.question,   // JSON string of comparison sequence
        answer: puzzleData.answer,       // JSON string of correct answers
        hint: puzzleData.hint,           // Instructions
        difficulty: difficulty,
        options: [], // No multiple choice options
        metadata: puzzleData.metadata    // Generation stats
      };
    }
  }

File: android/app/src/main/java/com/kreativekoala/riddleverse/MathComparisonGenerator.kt
package com.kreativekoala.riddleverse

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import kotlin.math.*
import kotlin.random.Random

data class ComparisonPuzzleResult(
    val success: Boolean,
    val puzzleData: ComparisonPuzzleData? = null,
    val message: String? = null
)

data class ComparisonPuzzleData(
    val question: String,
    val answer: String,
    val hint: String,
    val difficulty: String,
    val metadata: Map<String, Any>
)

data class ComparisonDifficultyConfig(
    val sequenceLength: Int,
    val numberRange: Pair<Int, Int>,
    val operations: List<String>,
    val allowDecimals: Boolean,
    val allowNegatives: Boolean,
    val timeLimit: Int,
    val scoring: ComparisonScoringConfig
)

data class ComparisonScoringConfig(
    val pointsPerCorrect: Int,
    val timeBonus: Boolean,
    val streakMultiplier: Double
)

class MathComparisonGenerator {
    private val TAG = "MathComparisonGen"
    private var debugMode = false

    fun setDebugMode(enabled: Boolean) {
        debugMode = enabled
    }

    private fun debugLog(message: String, type: String = "info") {
        if (debugMode) {
            val emoji = when (type) {
                "error" -> "❌"
                "success" -> "✅"
                "warning" -> "⚠️"
                else -> "🔄"
            }
            Log.d(TAG, "$emoji MATH_COMPARISON: $message")
        }
    }

    /**
     * Generate a complete mathematical comparison puzzle
     */
    fun generateMathComparisonPuzzle(difficulty: String = "easy"): ComparisonPuzzleResult {
        debugLog("🎯 Generating $difficulty mathematical comparison puzzle...")

        return try {
            val difficultyConfig = getDifficultyConfig(difficulty)
            val puzzleSequence = generatePuzzleSequence(difficultyConfig)

            val questionData = JSONObject().apply {
                put("sequence", JSONArray().apply {
                    puzzleSequence.forEach { pair ->
                        put(JSONObject().apply {
                            put("pairNumber", pair.pairNumber)
                            put("leftValue", pair.leftValue)
                            put("rightValue", pair.rightValue)
                            put("leftNumeric", pair.leftNumeric)
                            put("rightNumeric", pair.rightNumeric)
                            put("correctAnswer", pair.correctAnswer)
                            put("operationType", pair.operationType)
                            put("difficulty", pair.difficulty)
                        })
                    }
                })
                put("totalPairs", puzzleSequence.size)
                put("difficulty", difficulty)
                put("timeLimit", difficultyConfig.timeLimit)
                put("scoring", JSONObject().apply {
                    put("pointsPerCorrect", difficultyConfig.scoring.pointsPerCorrect)
                    put("timeBonus", difficultyConfig.scoring.timeBonus)
                    put("streakMultiplier", difficultyConfig.scoring.streakMultiplier)
                })
                put("instructions", "Compare the values and select which is greater, or select EQUAL if they are the same.")
                put("metadata", JSONObject().apply {
                    put("generatedAt", System.currentTimeMillis())
                    put("difficulty", difficulty)
                    put("sequenceLength", puzzleSequence.size)
                    put("operationTypes", JSONObject().apply {
                        getOperationTypes(puzzleSequence).forEach { (type, count) ->
                            put(type, count)
                        }
                    })
                })
            }

            val answerData = JSONObject().apply {
                put("correctAnswers", JSONArray().apply {
                    puzzleSequence.forEach { pair ->
                        put(pair.correctAnswer)
                    }
                })
                put("scoring", JSONObject().apply {
                    put("pointsPerCorrect", difficultyConfig.scoring.pointsPerCorrect)
                    put("timeBonus", difficultyConfig.scoring.timeBonus)
                    put("streakMultiplier", difficultyConfig.scoring.streakMultiplier)
                })
                put("maxScore", puzzleSequence.size * difficultyConfig.scoring.pointsPerCorrect)
            }

            val puzzleData = ComparisonPuzzleData(
                question = questionData.toString(),
                answer = answerData.toString(),
                hint = "Compare ${puzzleSequence.size} mathematical expressions and choose the greater value or EQUAL",
                difficulty = difficulty,
                metadata = mapOf(
                    "totalPairs" to puzzleSequence.size,
                    "generatedAt" to System.currentTimeMillis(),
                    "puzzleType" to "mathematical_comparison"
                )
            )

            debugLog("✅ Generated ${puzzleSequence.size} comparison pairs for $difficulty difficulty", "success")
            ComparisonPuzzleResult(success = true, puzzleData = puzzleData)

        } catch (error: Exception) {
            debugLog("❌ Generation failed: ${error.message}", "error")
            ComparisonPuzzleResult(
                success = false,
                message = "Math comparison generation failed: ${error.message}"
            )
        }
    }

    /**
     * Get difficulty configuration
     */
    private fun getDifficultyConfig(difficulty: String): ComparisonDifficultyConfig {
        return when (difficulty.lowercase()) {
            "easy" -> ComparisonDifficultyConfig(
                sequenceLength = 8,
                numberRange = Pair(1, 20),
                operations = listOf("addition", "subtraction", "simple_multiplication"),
                allowDecimals = false,
                allowNegatives = false,
                timeLimit = 60000, // 60 seconds
                scoring = ComparisonScoringConfig(
                    pointsPerCorrect = 100,
                    timeBonus = true,
                    streakMultiplier = 1.1
                )
            )
            "medium" -> ComparisonDifficultyConfig(
                sequenceLength = 12,
                numberRange = Pair(1, 50),
                operations = listOf("addition", "subtraction", "multiplication", "division", "mixed"),
                allowDecimals = true,
                allowNegatives = false,
                timeLimit = 90000, // 90 seconds
                scoring = ComparisonScoringConfig(
                    pointsPerCorrect = 150,
                    timeBonus = true,
                    streakMultiplier = 1.2
                )
            )
            "hard" -> ComparisonDifficultyConfig(
                sequenceLength = 15,
                numberRange = Pair(1, 100),
                operations = listOf("multiplication", "division", "mixed", "powers", "fractions"),
                allowDecimals = true,
                allowNegatives = true,
                timeLimit = 120000, // 120 seconds
                scoring = ComparisonScoringConfig(
                    pointsPerCorrect = 200,
                    timeBonus = true,
                    streakMultiplier = 1.3
                )
            )
            else -> getDifficultyConfig("medium") // Default fallback
        }
    }

    /**
     * Generate a sequence of comparison pairs with progressive difficulty
     */
    private fun generatePuzzleSequence(config: ComparisonDifficultyConfig): List<ComparisonPair> {
        val sequence = mutableListOf<ComparisonPair>()
        val sequenceLength = config.sequenceLength

        // Progressive difficulty within the sequence
        for (i in 0 until sequenceLength) {
            val progressRatio = i.toDouble() / sequenceLength
            val currentConfig = getProgressiveConfig(config, progressRatio)

            val pair = generateComparisonPair(currentConfig, i + 1)
            sequence.add(pair)
        }

        return sequence
    }

    /**
     * Adjust difficulty progressively through the sequence
     */
    private fun getProgressiveConfig(baseConfig: ComparisonDifficultyConfig, progressRatio: Double): ComparisonDifficultyConfig {
        val (min, max) = baseConfig.numberRange

        // Increase number range as we progress
        val rangeDiff = max - min
        val newMax = (min + rangeDiff * (0.3 + progressRatio * 0.7)).roundToInt()

        return baseConfig.copy(
            numberRange = Pair(min, newMax)
        )
    }

    /**
     * Generate a single comparison pair
     */
    private fun generateComparisonPair(config: ComparisonDifficultyConfig, pairNumber: Int): ComparisonPair {
        val complexity = (pairNumber - 1).toDouble() / config.sequenceLength

        // Choose operation type based on difficulty progression
        val operationType = selectOperation(config.operations, complexity)

        // Generate the two values to compare
        val leftValue = generateValue(operationType, config, "left")
        val rightValue = generateValue(operationType, config, "right")

        // Calculate actual numeric values
        val leftNumeric = evaluateExpression(leftValue)
        val rightNumeric = evaluateExpression(rightValue)

        // Determine correct answer
        val tolerance = 0.001 // For floating point comparison
        val correctAnswer = when {
            abs(leftNumeric - rightNumeric) < tolerance -> "equal"
            leftNumeric > rightNumeric -> "left"
            else -> "right"
        }

        return ComparisonPair(
            pairNumber = pairNumber,
            leftValue = leftValue,
            rightValue = rightValue,
            leftNumeric = leftNumeric,
            rightNumeric = rightNumeric,
            correctAnswer = correctAnswer,
            operationType = operationType,
            difficulty = calculatePairDifficulty(leftValue, rightValue, operationType)
        )
    }

    /**
     * Select operation type based on complexity
     */
    private fun selectOperation(availableOps: List<String>, complexity: Double): String {
        // Weight operations by complexity
        val operationWeights = mapOf(
            "addition" to 0.1,
            "subtraction" to 0.2,
            "simple_multiplication" to 0.3,
            "multiplication" to 0.5,
            "division" to 0.7,
            "mixed" to 0.8,
            "powers" to 0.9,
            "fractions" to 0.95
        )

        // Filter operations that are appropriate for current complexity
        val suitableOps = availableOps.filter { op ->
            (operationWeights[op] ?: 0.5) <= (complexity + 0.3)
        }

        return if (suitableOps.isNotEmpty()) {
            suitableOps.random()
        } else {
            availableOps.firstOrNull() ?: "addition" // Fallback
        }
    }

    /**
     * Generate a mathematical value/expression
     */
    private fun generateValue(operationType: String, config: ComparisonDifficultyConfig, side: String): String {
        val numberRange = config.numberRange
        val allowDecimals = config.allowDecimals

        return when (operationType) {
            "addition" -> generateAddition(numberRange, allowDecimals)
            "subtraction" -> generateSubtraction(numberRange, allowDecimals)
            "simple_multiplication" -> generateSimpleMultiplication(numberRange)
            "multiplication" -> generateMultiplication(numberRange, allowDecimals)
            "division" -> generateDivision(numberRange, allowDecimals)
            "mixed" -> generateMixedExpression(numberRange, allowDecimals)
            "powers" -> generatePowerExpression(numberRange)
            "fractions" -> generateFractionExpression(numberRange)
            else -> generateSimpleNumber(numberRange, allowDecimals)
        }
    }

    /**
     * Generate addition expression
     */
    private fun generateAddition(range: Pair<Int, Int>, allowDecimals: Boolean): String {
        val a = randomNumber(range, allowDecimals)
        val b = randomNumber(range, allowDecimals)
        return "$a + $b"
    }

    /**
     * Generate subtraction expression
     */
    private fun generateSubtraction(range: Pair<Int, Int>, allowDecimals: Boolean): String {
        val a = randomNumber(range, allowDecimals)
        val maxB = minOf(a.toInt(), range.second)
        val b = randomNumber(Pair(range.first, maxB), allowDecimals)
        return "$a - $b"
    }

    /**
     * Generate simple multiplication (single digits)
     */
    private fun generateSimpleMultiplication(range: Pair<Int, Int>): String {
        val a = randomNumber(Pair(2, 9), false).toInt()
        val b = randomNumber(Pair(2, 9), false).toInt()
        return "$a × $b"
    }

    /**
     * Generate multiplication expression
     */
    private fun generateMultiplication(range: Pair<Int, Int>, allowDecimals: Boolean): String {
        // Keep numbers smaller for multiplication
        val maxVal = minOf(range.second, 15)
        val a = randomNumber(Pair(range.first, maxVal), allowDecimals)
        val b = randomNumber(Pair(range.first, maxVal), allowDecimals)
        return "$a × $b"
    }

    /**
     * Generate division expression
     */
    private fun generateDivision(range: Pair<Int, Int>, allowDecimals: Boolean): String {
        val b = randomNumber(Pair(2, minOf(range.second, 10)), false).toInt()
        val result = randomNumber(Pair(1, range.second), allowDecimals)
        val a = (b * result.toDouble()).let {
            if (allowDecimals) it else it.toInt().toDouble()
        }
        return "$a ÷ $b"
    }

    /**
     * Generate mixed expression
     */
    private fun generateMixedExpression(range: Pair<Int, Int>, allowDecimals: Boolean): String {
        val operations = listOf("+", "-", "×")
        val op1 = operations.random()
        val op2 = operations.random()

        val maxVal = minOf(range.second, 12)
        val a = randomNumber(Pair(range.first, maxVal), allowDecimals)
        val b = randomNumber(Pair(range.first, maxVal), allowDecimals)
        val c = randomNumber(Pair(range.first, maxVal), allowDecimals)

        return "$a $op1 $b $op2 $c"
    }

    /**
     * Generate power expression
     */
    private fun generatePowerExpression(range: Pair<Int, Int>): String {
        val base = randomNumber(Pair(2, minOf(range.second, 8)), false).toInt()
        val exponent = randomNumber(Pair(2, 4), false).toInt()
        return "$base^$exponent"
    }

    /**
     * Generate fraction expression
     */
    private fun generateFractionExpression(range: Pair<Int, Int>): String {
        val numerator = randomNumber(Pair(1, range.second), false).toInt()
        val denominator = randomNumber(Pair(2, 10), false).toInt()
        return "$numerator/$denominator"
    }

    /**
     * Generate simple number
     */
    private fun generateSimpleNumber(range: Pair<Int, Int>, allowDecimals: Boolean): String {
        return randomNumber(range, allowDecimals).toString()
    }

    /**
     * Generate random number within range
     */
    private fun randomNumber(range: Pair<Int, Int>, allowDecimals: Boolean): Double {
        val min = range.first
        val max = range.second

        return if (allowDecimals && Random.nextDouble() < 0.3) { // 30% chance for decimals
            ((Random.nextDouble() * (max - min) + min) * 10).roundToInt() / 10.0
        } else {
            Random.nextInt(min, max + 1).toDouble()
        }
    }

    /**
     * Evaluate mathematical expression to numeric value
     */
    private fun evaluateExpression(expression: String): Double {
        return try {
            // Convert mathematical symbols and evaluate
            val result = when {
                // Handle fractions
                expression.contains("/") && !expression.contains(" ") -> {
                    val parts = expression.split("/")
                    if (parts.size == 2) {
                        val num = parts[0].toDouble()
                        val den = parts[1].toDouble()
                        num / den
                    } else {
                        evaluateComplexExpression(expression)
                    }
                }
                // Handle powers
                expression.contains("^") -> {
                    val parts = expression.split("^")
                    if (parts.size == 2) {
                        val base = parts[0].toDouble()
                        val exp = parts[1].toDouble()
                        base.pow(exp)
                    } else {
                        evaluateComplexExpression(expression)
                    }
                }
                // Handle simple operations
                else -> evaluateComplexExpression(expression)
            }
            result
        } catch (error: Exception) {
            debugLog("Error evaluating expression \"$expression\": ${error.message}", "error")
            0.0
        }
    }

    /**
     * Evaluate complex mathematical expressions
     */
    private fun evaluateComplexExpression(expression: String): Double {
        // Replace mathematical symbols
        val jsExpression = expression
            .replace("×", "*")
            .replace("÷", "/")
            .replace("^", "**")

        // Simple expression evaluator (handles basic arithmetic with order of operations)
        return evaluateArithmetic(jsExpression)
    }

    /**
     * Simple arithmetic evaluator
     */
    private fun evaluateArithmetic(expression: String): Double {
        // Remove spaces
        val expr = expression.replace(" ", "")

        // Handle parentheses first (if any)
        var workingExpr = expr

        // Handle multiplication and division first
        workingExpr = processOperations(workingExpr, listOf("*", "/"))

        // Handle addition and subtraction
        workingExpr = processOperations(workingExpr, listOf("+", "-"))

        return workingExpr.toDoubleOrNull() ?: 0.0
    }

    /**
     * Process specific operations in order
     */
    private fun processOperations(expression: String, operations: List<String>): String {
        var expr = expression

        for (op in operations) {
            while (expr.contains(op)) {
                val regex = when (op) {
                    "*" -> """(\d+(?:\.\d+)?)\*(\d+(?:\.\d+)?)""".toRegex()
                    "/" -> """(\d+(?:\.\d+)?)/(\d+(?:\.\d+)?)""".toRegex()
                    "+" -> """(\d+(?:\.\d+)?)\+(\d+(?:\.\d+)?)""".toRegex()
                    "-" -> """(\d+(?:\.\d+)?)-(\d+(?:\.\d+)?)""".toRegex()
                    else -> continue
                }

                val match = regex.find(expr)
                if (match != null) {
                    val a = match.groupValues[1].toDouble()
                    val b = match.groupValues[2].toDouble()
                    val result = when (op) {
                        "*" -> a * b
                        "/" -> if (b != 0.0) a / b else 0.0
                        "+" -> a + b
                        "-" -> a - b
                        else -> 0.0
                    }
                    expr = expr.replaceFirst(match.value, result.toString())
                } else {
                    break
                }
            }
        }

        return expr
    }

    /**
     * Calculate difficulty rating for a pair
     */
    private fun calculatePairDifficulty(leftValue: String, rightValue: String, operationType: String): Double {
        var difficulty = 1.0

        // Base difficulty by operation type
        val operationDifficulty = mapOf(
            "addition" to 1.0,
            "subtraction" to 1.2,
            "simple_multiplication" to 1.5,
            "multiplication" to 2.0,
            "division" to 2.5,
            "mixed" to 3.0,
            "powers" to 3.5,
            "fractions" to 4.0
        )

        difficulty *= operationDifficulty[operationType] ?: 1.0

        // Increase difficulty for longer expressions
        val avgLength = (leftValue.length + rightValue.length) / 2.0
        difficulty *= (1 + avgLength * 0.1)

        return (difficulty * 10).roundToInt() / 10.0 // Round to 1 decimal
    }

    /**
     * Get summary of operation types used
     */
    private fun getOperationTypes(sequence: List<ComparisonPair>): Map<String, Int> {
        val types = mutableMapOf<String, Int>()
        sequence.forEach { pair ->
            types[pair.operationType] = (types[pair.operationType] ?: 0) + 1
        }
        return types
    }
}